keras/utils/data_utils.py

```diff
@@ -282,3 +295,314 @@ def validate_file(fpath, file_hash, algorithm='auto', chunk_size=65535):
+    enqueuer.start()
+    datas = enqueuer.get()
+    for data in datas:
+        # Use the inputs; training, evaluating, predicting.
+        # ... stop sometime.
+    enqueuer.close()
+    ```
+
+    The `enqueuer.get()` should be an infinite stream of datas.
+
+    """
+
+    @abstractmethod
+    def is_running(self):
+        raise NotImplemented
+
+    @abstractmethod
+    def start(self, workers=1, max_q_size=10):
+        """Start the handler's workers.
+
+        # Arguments
+            workers: number of worker threads
+            max_q_size: queue size (when full, threads could block on put())
+        """
+        raise NotImplemented
+
+    @abstractmethod
+    def stop(self, timeout=None):
+        """Stop running threads and wait for them to exit, if necessary.
+
+        Should be called by the same thread which called start().
+
+        # Arguments
+            timeout: maximum time to wait on thread.join()
+        """
+        raise NotImplemented
+
+    @abstractmethod
+    def get(self):
+        """Create a generator to extract data from the queue. Skip the data if it's None.
+
+        #Returns
+            A generator
+        """
+        raise NotImplemented
+
+
+class OrderedEnqueuer(DatasetEnqueuer):
+    """Builds a Enqueuer from a Dataset.
+
+    Used in `fit_generator`, `evaluate_generator`, `predict_generator`.
+
+    # Arguments
+        dataset: A `keras.data.dataset.Dataset` object.
+        pickle_safe: use multiprocessing if True, otherwise threading
+        scheduling: Sequential querying of datas if 'sequential', random otherwise.
+    """
+
+    def __init__(self, dataset, pickle_safe=False, scheduling='sequential'):
+        self.dataset = dataset
+        self.pickle_safe = pickle_safe
+        self.scheduling = scheduling
+        self.workers = 0
+        self.executor = None
+        self.queue = None
+        self.run_thread = None
+        self.stop_signal = None
+
+    def is_running(self):
+        return self.stop_signal is not None and not self.stop_signal.is_set()
+
+    def start(self, workers=1, max_q_size=10):
+        """Start the handler's workers.
+
+        # Arguments
+            workers: number of worker threads
+            max_q_size: queue size (when full, workers could block on put())
+        """
+        if self.pickle_safe:
+            self.executor = multiprocessing.Pool(workers)
+        else:
+            self.executor = ThreadPool(workers)
+        self.queue = queue.Queue(max_q_size)
+        self.stop_signal = threading.Event()
+        self.run_thread = threading.Thread(target=self._run)
+        self.run_thread.daemon = True
+        self.run_thread.start()
+
+    def _run(self):
+        """ Function to submit request to the executor and queue the `Future` objects."""
+        indexes = list(range(len(self.dataset)))
+        while True:
+            if self.scheduling is not 'sequential':
+                random.shuffle(indexes)
+            for i in indexes:
```

Comment:
I just realized that I missed something important. 
These `indexes` are not the indexes of individual samples.
These are indexes of batches.
It means that forever you will have the same samples in the same batch, even if you shuffle the indexes.
It takes the edge out of some important operations (i.e. batch-normalization).

I find this a bit confusing.

What I would expect is that the Dataset will store samples, not batches, and that `getitem` would return a single sample. I don't think that a Dataset should know about the `batch_size`. It makes more sense to me that the packaging of the samples into batches would happen on the fly, in the `DatasetEnqueuer.get()` method.  If the `batch_size` is 32, then Enqueuer should pop out 32 elements from the queue (or less if there is a StopIteration), and call a `batcher` function (perhaps it can be provided by `Dataset`) that turns a list of samples into a batch.